<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Differential Expression Analysis of expressed genes in Prostate Cancer</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="readme_files/libs/clipboard/clipboard.min.js"></script>
<script src="readme_files/libs/quarto-html/quarto.js"></script>
<script src="readme_files/libs/quarto-html/popper.min.js"></script>
<script src="readme_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="readme_files/libs/quarto-html/anchor.min.js"></script>
<link href="readme_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="readme_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="readme_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="readme_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="readme_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Differential Expression Analysis of expressed genes in Prostate Cancer</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="the-aim" class="level1">
<h1>THE AIM</h1>
<p>Unlike the current methods, the thesis aims to help detect new genetic biomarkers of Prostate Cancer. For this purpose, I used multiple approaches beginning with differential expression analysis to focus on the statistically significant different genes. those genes were used in advanced steps to gain biological insights by enrichment analysis and network analysis and to prioritize the most important genes some classification models were trained to know which are the most important genes that makes the cell cancerous.</p>
<p>But this READme aims to show exactly the steps I followed to conduct my analysis regarding Data collection, filtration exploration, and analysis including differential expression analysis and machine learning. in addition some comments on the methods I used.</p>
</section>
<section id="data-collection" class="level1">
<h1>Data Collection</h1>
<p>The gene expression data utilized in this study was sourced from the GDC Data Portal and imported directly into our development environment using the TCGAbiolinks R<br>
package. This method was way easier and faster to get your data with the desired features but it maybe computationaly consuming to do it in case of big samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="do">##building query data</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>gdc.query <span class="ot">=</span> <span class="fu">GDCquery</span>(</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">project =</span> <span class="st">"TCGA-PRAD"</span> ,</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data.category =</span> <span class="st">'Transcriptome Profiling'</span> ,</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">experimental.strategy =</span> <span class="st">'RNA-Seq'</span>,</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">workflow.type =</span> <span class="st">'STAR - Counts'</span>,</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">access =</span> <span class="st">'open'</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>results<span class="ot">&lt;-</span> <span class="fu">getResults</span>(gdc.query)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>results</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we download the data and retrieve the count matrix and the metadata.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#downlaod the data </span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">GDCdownload</span>(gdc.query)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co">#prepare the data</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>TCGA.data <span class="ot">&lt;-</span> <span class="fu">GDCprepare</span>(gdc.query, <span class="at">summarizedExperiment =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>data.matrix <span class="ot">&lt;-</span> <span class="fu">assay</span>(TCGA.data, <span class="st">'unstranded'</span>)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>rowData <span class="ot">&lt;-</span> <span class="fu">rowData</span>(TCGA.data)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>colData <span class="ot">&lt;-</span>  <span class="fu">colData</span>(TCGA.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This data originated from the TCGA-Prostate Adenocarcinoma (PRAD) project and had already undergone essential preprocessing steps, converting raw sequencing data from Fastq files into raw count expression data. We opted for the RNA-Seq experimental strategy and chose the STAR-Counts workflow type, specifically focusing on “unstranded” read count analysis that considers both sense and antisense transcripts. Access to this data was granted through an open-access license, ensuring its availability for research purposes. The dataset comprised 553 cases, representing samples from individuals diagnosed with prostate adenocarcinoma (PRAD). Importantly, the data were obtained in their raw form without normalization or transformation, allowing for a thorough analysis of the gene expression values and metadata associated with the samples.</p>
<section id="filtration-and-normalization" class="level2">
<h2 class="anchored" data-anchor-id="filtration-and-normalization">Filtration and Normalization</h2>
<p><br>
Before normalization, RNAs with near-zero or zero read count values across all samples<br>
were excluded from the study. Additionally, a filtration step based on protein-coding<br>
information was applied to focus on actively transcribed genes relevant to the functional<br>
aspects of the transcriptome. This filtration ensured the analysis centered on biologically<br>
relevant genes with known protein-coding potential. Following the filtration process for<br>
protein-coding genes, a refined Gene Expression Matrix comprising 18,611 protein-coding genes was constructed. This filtered dataset served as the basis for our exploratory analysis. Subsequently, the data was normalized using the variance-stabilizing transform (VST) through the DESeq2 R-project package. VST was employed to accommodate the different variances between RNAs, as it diminishes the dependency of variance on the mean value of the read counts. This normalization strategy aimed to reduce systematic errors and enhance the reliability of our subsequent exploratory analysis.</p>
</section>
<section id="exploratory-data-analysis" class="level2">
<h2 class="anchored" data-anchor-id="exploratory-data-analysis">Exploratory Data Analysis</h2>
<p>Following the essential data preprocessing steps, the dataset obtained from TCGA was explored through visualization techniques, including Principal Component Analysis (PCA), density plots, and box plots. The subsequentt section elaborates on the outcomes and insights from each visualization method.</p>
<p>The study’s initial phase, known as exploratory analysis, aims to run the data through several visualization and filtration processes for three reasons: to find RNAs that might not be relevant to our investigation; to find samples that might be incorrectly labeled, misplaced, or otherwise flawed and need to be excluded from the analysis; and, finally, to confirm that the available tumor and non-tumor samples show a distinct and obvious difference. These goals will mostly be achieved through data visualization, making the patterns displayed by the samples and RNAs easy to see.</p>
<section id="density-plot" class="level3">
<h3 class="anchored" data-anchor-id="density-plot">Density Plot</h3>
<p>In this section, the log-transformed counts were used to produce density plots. Upon examining the density plots for both tumor and normal samples, a consistent distribution of expression values was observed across all samples in both categories. This uniformity suggests the absence of outliers or abnormalities within the sample set. This finding underscores the reliability and consistency of the data set, laying a strong foundation for further analysis and interpretation.</p>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="images/tumor%20density.png" class="img-fluid figure-img" width="1000"></p>
<figcaption class="figure-caption">Tumor samples density plots</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/normal%20density.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Normal samples density plot</figcaption>
</figure>
</div>
</section>
</section>
<section id="box-plots" class="level2">
<h2 class="anchored" data-anchor-id="box-plots">Box Plots</h2>
<p>The utilization of box plots to visualize raw expression values showcased high comparability among the samples before normalization. Notably, these plots did not exhibit any peculiar or abnormal distributions within any of the samples. This uniformity across the dataset, even before normalization, signifies the robustness of the data and bolsters confidence in the subsequent analytical procedures.</p>
</section>
<section id="pca" class="level2">
<h2 class="anchored" data-anchor-id="pca">PCA</h2>
<p>Principal Component Analysis (PCA) was employed on the DESeqTransform object, which contains normalized data processed using the DESeq2 package’s VTS method. PCA revealed distinct clustering between tumor and normal samples with some overlap. This overlap may be attributed to prostate cancer’s inherent heterogeneity and the intricate transcriptional patterns associated with it.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/PCAnormalized.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">PCA of normalized counts</figcaption>
</figure>
</div>
</section>
</section>
<section id="differential-expression-analysis" class="level1">
<h1>Differential Expression Analysis</h1>
<p>Differential analysis was conducted using the DESeq2 package with stringent criteria<br>
(|log2 FoldChange| &gt; 1.5 and padj &lt; 0.01). to identify genes with significant expression<br>
changes in prostate adenocarcinoma (PRAD). A total of 1224 differentially expressed<br>
genes (DEGs) were identified, with 675 upregulated and 549 downregulated in tumor<br>
samples compared to normal samples. A volcano plot was constructed to visualize these<br>
changes and identify the top genes based on adjusted P and log2 fold change values<br>
for further investigation. A volcano plot was constructed to visualize these changes<br>
and identify the top genes based on adjusted P and log2 fold change values for further<br>
investigation.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/f.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Volcano Plot reveals the upregulated and downregulated genes</figcaption>
</figure>
</div>
<section id="heat-map" class="level2">
<h2 class="anchored" data-anchor-id="heat-map">Heat map</h2>
<p>Subsequently, a heatmap was generated to depict gene expression levels across samples, focusing on the top 100 DEGs. The heatmap highlighted the heterogeneity of<br>
tumor samples compared to the more consistent transcriptional profiles of normal sam-<br>
ples within this gene subset. Notably, the heatmap illustrated the downregulation of<br>
the first 13 DEGs in normal samples, while the remaining 87 genes were predominantly<br>
upregulated in most normal samples.</p>
<p><img src="images/Picture1.png" class="img-fluid" alt="Heatmap of Top 100 DEGs"><br>
</p>
</section>
<section id="pca-for-differentially-expressed-genes" class="level2">
<h2 class="anchored" data-anchor-id="pca-for-differentially-expressed-genes">PCA for Differentially Expressed Genes</h2>
<p><br>
To delve deeper into the transcriptomic profiles and visualize patterns and variations<br>
within the dataset, we conducted principal component analysis (PCA) on the expression<br>
data. We generated both 2D and 3D PCA plots. The PCA analysis yielded clear segregation between tumor and normal samples in the 2D and 3D plots. This observation underscores the robustness of our findings in discriminating between normal and cancerous tissues in the context of prostate adenocarcinoma<br>
(PRAD). Notably, this enhanced separation in the PCA plots with selected genes suggests improved discriminative power compared to PCA using all genes, further validating<br>
the efficacy of our analysis.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/PCA2D.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">PCA of DEGs</figcaption>
</figure>
</div>
</section>
<section id="machine-learning" class="level2">
<h2 class="anchored" data-anchor-id="machine-learning">Machine Learning</h2>
<p>The machine learning part of the thesis aims to prioritize the most important genes differentiating normal and tumor tissues by training accurate classification models to shed light on the most important variables in the data.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/ml.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">machine learning workflow</figcaption>
</figure>
</div>
<section id="data-pre-processing" class="level3">
<h3 class="anchored" data-anchor-id="data-pre-processing">1. Data Pre-processing</h3>
<p>The process begins with acquiring gene expression data from cancer and normal samples. This raw data undergoes pre-processing steps such as normalization, imputation of missing values, and transformation to ensure consistency and reliability. Our data set is the DEGs expression values, which are already preprocessed.</p>
</section>
<section id="feature-selection" class="level3">
<h3 class="anchored" data-anchor-id="feature-selection">2. Feature selection</h3>
<p>In the realm of biomarker discovery, identifying the most relevant features (genes) is crucial to reduce dimensionality and retain only the informative features. This step is done by obtaining DEGs between samples.</p>
</section>
<section id="data-splitting" class="level3">
<h3 class="anchored" data-anchor-id="data-splitting">3. Data Splitting</h3>
<p>The dataset is divided into training 70%, testing 20%, and validation 10% sets. The training set is used to train the machine learning models, the testing set evaluates model performance during development, and the validation set assesses the model’s generalization on unseen data.</p>
</section>
<section id="model-selection" class="level3">
<h3 class="anchored" data-anchor-id="model-selection">4. Model Selection</h3>
<p>we chose 4 clustering algorithms for supervised learning in order to differentiate between samples depending on their transcription profiles of DEGs.</p>
<section id="svm-support-vector-machines" class="level4">
<h4 class="anchored" data-anchor-id="svm-support-vector-machines">SVM (Support Vector Machines):</h4>
<p>SVM is typically used for classification tasks that aim to find a hyperplane that best separates different classes. In clustering, SVM can be employed in a way known as one-class SVM, which essentially tries to isolate one class of data from the rest. This can be useful in outlier detection or anomaly detection within clusters.</p>
</section>
<section id="rf-random-forest" class="level4">
<h4 class="anchored" data-anchor-id="rf-random-forest">RF (Random Forest):</h4>
<p>RF is an ensemble learning method that constructs multiple decision trees during training and outputs the mode of the classes (for classification) or the average prediction (for regression). In clustering, RF can be used indirectly by training it on the data and then using the leaf node assignments of the trees as cluster labels. This approach is known as “using RF for clustering via leaf node assignments.”</p>
</section>
<section id="xgboost" class="level4">
<h4 class="anchored" data-anchor-id="xgboost">XGBoost:</h4>
<p>XGBoost is another ensemble learning technique that is particularly powerful for classification and regression tasks. Like RF, XGBoost can be indirectly used for clustering by training it on the data and then extracting cluster assignments based on the leaf nodes of the boosted trees.</p>
</section>
<section id="neural-networks" class="level4">
<h4 class="anchored" data-anchor-id="neural-networks">Neural Networks:</h4>
<p>The usage of neural networks</p>
</section>
</section>
<section id="model-training" class="level3">
<h3 class="anchored" data-anchor-id="model-training">5. Model Training</h3>
<p>Various machine learning algorithms, such as Support Vector Machines (SVM), Random Forests, extreme Gradient Boosting (XGB), and Neural Networks, are trained using the training data set. These models learn patterns and relationships within the data to accurately classify cancer and normal samples.</p>
</section>
<section id="model-evaluation" class="level3">
<h3 class="anchored" data-anchor-id="model-evaluation">6. Model Evaluation</h3>
<p>The trained models are evaluated using the testing dataset to assess their performance metrics such as accuracy, precision, recall, and F1 score. This evaluation ensures the models generalize well and can reliably classify unseen samples.</p>
</section>
<section id="feature-importance" class="level3">
<h3 class="anchored" data-anchor-id="feature-importance">7. Feature Importance</h3>
<p>Post-training models provide insights into feature importance, highlighting the genes (variables) that contribute significantly to the classification task. This step is pivotal in biomarker discovery, as it identifies potential biomarkers for further validation and clinical translation.</p>
</section>
</section>
<section id="model-evaluation-1" class="level2">
<h2 class="anchored" data-anchor-id="model-evaluation-1">Model Evaluation</h2>
<p>The trained models underwent evaluation post-validation, assessing their performance through diverse metrics based on predictions made on the test set. we estimated 4 metrics</p>
<p><img src="images/metrics.png" class="img-fluid"></p>
<p>As depicted in the figure, all models exhibited notably high accuracy, underscoring<br>
their robust performance. After confirming their high predictive accuracy, we extracted<br>
the most influential variables, specifically the genes pivotal in sample classification. This<br>
iterative process of model evaluation, validation, and feature selection validates our pre-<br>
dictive models’ efficacy and unveils key genetic determinants critical for accurate sample<br>
classification. The most important variables in each model have been obtained to gain<br>
insights about what could be the possible biomarkers.</p>
</section>
<section id="section" class="level2">
<h2 class="anchored" data-anchor-id="section"></h2>
</section>
<section id="section-1" class="level2">
<h2 class="anchored" data-anchor-id="section-1"></h2>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>