# Integrative Analysis of Differentialy Expressed genes in Prostate Cancer

## Abstract:

Prostate cancer is a complex and heterogeneous disease with diverse molecular subtypes and clinical outcomes. In this thesis, we conducted a comprehensive analysis of prostate cancer using integrative omics approaches to unravel key biomarkers and pathways associated with disease progression.
The study utilized RNA-seq data from 553 samples and machine learning algorithms to identify 1224 differentially expressed genes (DEGs) between normal and tumor samples. Gene set enrichment analysis (GSEA) revealed significant associations with biological processes such as serine-type endopeptidase activity regulation and pathways linked to cancer growth and metastasis.
Machine learning models demonstrated high accuracy in predicting sample classification, with top-ranked genes including EPHA10, ABCG2, and GSTM3 identified as potential biomarkers for prostate cancer. Network analysis highlighted gene clusters and interactions, further validating our findings through publication enrichment and biological process enrichment.
Our thesis proposes strategies for experimental validation, multi-omics integration, and clinical translation to advance precision medicine in prostate cancer diagnosis and treatment. The findings contribute to a deeper understanding of prostate cancer biology and hold promise for personalized therapeutic interventions and improved patient outcomes.

